libkbfs/folder_block_ops.go

```diff
@@ -803,39 +816,237 @@ func (fbo *folderBlockOps) GetDir(
+	// In case it was removed in the cache but not flushed yet.
+	delete(cacheEntry.dels, newName)
+	fbo.deCache[dir.tailPointer().Ref()] = cacheEntry
+}
+
+// AddDirEntryInCache adds a brand new entry to the given directory in
+// the cache, which will get applied to the dirty block on subsequent
+// fetches for the directory.  The new entry must not yet have a cache
+// entry itself.
+func (fbo *folderBlockOps) AddDirEntryInCache(lState *lockState, dir path,
+	newName string, newDe DirEntry) {
+	fbo.blockLock.Lock(lState)
+	defer fbo.blockLock.Unlock(lState)
+	fbo.addDirEntryInCacheLocked(lState, dir, newName, newDe)
+	// Add target dir entry as well.
+	if newDe.IsInitialized() {
+		cacheEntry, ok := fbo.deCache[newDe.Ref()]
+		if ok {
+			panic("New entry shouldn't already exist")
+		}
+		cacheEntry.dirEntry = newDe
+		fbo.deCache[newDe.Ref()] = cacheEntry
+	}
+}
+
+func (fbo *folderBlockOps) removeDirEntryInCacheLocked(lState *lockState,
+	dir path, oldName string) {
+	fbo.blockLock.AssertLocked(lState)
+	cacheEntry := fbo.deCache[dir.tailPointer().Ref()]
+	if cacheEntry.dels == nil {
+		cacheEntry.dels = make(map[string]bool)
+	}
+	cacheEntry.dels[oldName] = true
+	// In case it was added in the cache but not flushed yet.
+	delete(cacheEntry.adds, oldName)
+	fbo.deCache[dir.tailPointer().Ref()] = cacheEntry
+}
+
+// RemoveDirEntryInCache removes an entry from the given directory in
+// the cache, which will get applied to the dirty block on subsequent
+// fetches for the directory.
+func (fbo *folderBlockOps) RemoveDirEntryInCache(lState *lockState, dir path,
+	oldName string) {
+	fbo.blockLock.Lock(lState)
+	defer fbo.blockLock.Unlock(lState)
+	fbo.removeDirEntryInCacheLocked(lState, dir, oldName)
+}
+
+// RenameDirEntryInCache updates the entries of both the old and new
+// parent dirs for the given target dir atomically (with respect to
+// blockLock).  It also updates the cache entry for the target, which
+// would have its Ctime changed. The updates will get applied to the
+// dirty blocks on subsequent fetches.
+//
+// The returned bool indicates whether or not the caller should clean
+// up the target cache entry when the effects of the operation are no
+// longer needed.
+func (fbo *folderBlockOps) RenameDirEntryInCache(lState *lockState,
+	oldParent path, oldName string, newParent path, newName string,
+	newDe DirEntry) (deleteTargetDirEntry bool) {
+	fbo.blockLock.Lock(lState)
+	defer fbo.blockLock.Unlock(lState)
+	fbo.addDirEntryInCacheLocked(lState, newParent, newName, newDe)
+	fbo.removeDirEntryInCacheLocked(lState, oldParent, oldName)
+	// If there's already an entry for the target, only update the
+	// Ctime on a rename.
+	cacheEntry, ok := fbo.deCache[newDe.Ref()]
+	if ok && cacheEntry.dirEntry.IsInitialized() {
+		cacheEntry.dirEntry.Ctime = newDe.Ctime
+	} else {
+		cacheEntry.dirEntry = newDe
+		fbo.deCache[newDe.Ref()] = cacheEntry
+		deleteTargetDirEntry = true
+	}
+	fbo.deCache[newDe.Ref()] = cacheEntry
```

Comment:
is this on purpose? seems to be the same as in line 898, so one of the lines must be redundant...